### PR TITLE
feat: no-reply/no-match improvements (VF-1997)

### PR DESCRIPTION
**Pull request by z4o4z**
**Fixes or implements VF-1997**

### Related PRs
https://github.com/voiceflow/libs/pull/115

### CODE REVIEW

File: lib/services/runtime/handlers/noReply.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    runtime.storage.set(StorageType.NO_REPLIES_COUNTER, noReplyCounter + 1);
+
+    runtime.trace.addTrace<Trace.PathTrace>({
+      type: BaseNode.Utils.TraceType.PATH,
+      payload: { path: 'reprompt' },
+    });
+
+    const output = node.noReply?.randomize
+      ? _.sample<string | Text.SlateTextValue>(nonEmptyNoReplies)
+      : nonEmptyNoReplies?.[runtime.storage.get<NoMatchCounterStorage>(StorageType.NO_REPLIES_COUNTER)! - 1];
```
Comment (DecathectZero) [inline lib/services/runtime/handlers/noReply.ts:53]:
```suggestion
      : nonEmptyNoReplies?.[noReplyCounter];
```

move line 44 (increment) after this to make it easier to read

**Review by DecathectZero — CHANGES_REQUESTED** [review]
`request: null` might not be the best implementation here, because we've essentially tied `null` to the launch function. Like you mentioned earlier:

https://github.com/voiceflow/general-runtime/blob/a8da9c94b2fcb72687d6ad8fe318c56c43c762e8/runtime/lib/Runtime/index.ts#L148

When you start the prototype tool, it sends a `request: null`, and will ALWAYS trigger the no reply logic. The only alternative I've found to this is changing the line above to just `this.setAction(Action.REQUEST)` and in the start block doing: 
Screen Shot 2021-11-23 at 9 37 54 PM

Although I do worry about the side effects.

**Comment by DecathectZero** [comment]
I'm hoping this fixes a lot of the actions/request/response issues:
https://github.com/voiceflow/general-runtime/pull/228

File: lib/services/runtime/utils.ts [hunk 2]
```diff
@@ -139,11 +125,14 @@ export const addButtonsIfExists = <N extends Request.NodeButton>(node: N, runtim
   }
 };
 
-export const getReadableConfidence = (confidence?: number) => ((confidence ?? 1) * 100).toFixed(2);
+export const getReadableConfidence = (confidence?: number): string => ((confidence ?? 1) * 100).toFixed(2);
 
 export const slateToPlaintext = (content: Text.SlateTextValue = []): string =>
   content.reduce<string>((acc, node) => acc + (SlateText.isText(node) ? node.text : slateToPlaintext(node.children)), '');
 
+export const removeEmptyPrompts = (prompts: Array<Text.SlateTextValue | string>): Array<Text.SlateTextValue | string> =>
```
Comment (DecathectZero) [inline lib/services/runtime/utils.ts:133]: shouldn't this mostly be done on the service?

File: lib/services/runtime/handlers/noMatch.ts [hunk 3]
```diff
@@ -1,57 +1,64 @@
...
-  ) as T | undefined;
+const convertDeprecatedNoMatch = ({ noMatch, elseId, noMatches, randomize, ...node }: NoMatchNode) =>
+  ({
+    noMatch: {
+      prompts: noMatch?.prompts ?? noMatches,
+      randomize: noMatch?.randomize ?? randomize,
+      nodeID: noMatch?.nodeID ?? elseId,
+    },
+    ...node,
+  } as NoMatchNode);
```
Comment (DecathectZero) [inline lib/services/runtime/handlers/noMatch.ts:26]: this is the one hack I have to do....the union type on `NoMatchNode` gets mad :(

File: lib/services/runtime/utils.ts (hunk 2 above)
Comment (z4o4z) [reply lib/services/runtime/utils.ts:133]: we have it in the service too, but on runtime, we are injecting variables, and some prompts can be empty due to empty strings in vars

**Review by DecathectZero — APPROVED** [review]

File: lib/services/runtime/handlers/noMatch.ts [hunk 4]
```diff
@@ -1,61 +1,71 @@
...
-    });
+export const NoMatchHandler = (utils: typeof utilsObj) => ({
+  handle: (_node: NoMatchNode, runtime: Runtime, variables: Store) => {
+    const node = convertDeprecatedNoMatch(_node);
+
+    const nonEmptyNoMatches = removeEmptyNoMatches(node);
+
+    const noMatchCounter = runtime.storage.get<NoMatchCounterStorage>(StorageType.NO_MATCHES_COUNTER) ?? 0;
+
+    if (noMatchCounter >= nonEmptyNoMatches.length) {
```
Comment (DecathectZero) [inline lib/services/runtime/handlers/noMatch.ts:44]:
```suggestion
    if (noMatchCounter > nonEmptyNoMatches.length) {
```

File: lib/services/runtime/handlers/noMatch.ts [hunk 5]
```diff
@@ -1,61 +1,71 @@
...
 
-      draft[StorageType.NO_MATCHES_COUNTER] = counter ? counter + 1 : 1;
-    });
+export const NoMatchHandler = (utils: typeof utilsObj) => ({
+  handle: (_node: NoMatchNode, runtime: Runtime, variables: Store) => {
+    const node = convertDeprecatedNoMatch(_node);
+
+    const nonEmptyNoMatches = removeEmptyNoMatches(node);
+
+    const noMatchCounter = runtime.storage.get<NoMatchCounterStorage>(StorageType.NO_MATCHES_COUNTER) ?? 0;
```
Comment (DecathectZero) [inline lib/services/runtime/handlers/noMatch.ts:42]:
```suggestion
    const noMatchCounter = runtime.storage.get<NoMatchCounterStorage>(StorageType.NO_MATCHES_COUNTER) ?? 1;
```

**Pull request merged**